### PR TITLE
feat(frontend): dapp modal

### PR DESCRIPTION
**Pull request by loki344**
# Motivation

Implement a modal to show the details of a dapp.

# Changes

- implement DappModal
- add modal to store
- enhance external link with class so that it can be better reused (IOC)

mobile with screenshot
image
mobile without screenshot
image
desktop with screenshot
image
desktop without screenshot
image

### CODE REVIEW

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	import ExternalLinkIcon from '$lib/components/ui/ExternalLinkIcon.svelte';
+	import Img from '$lib/components/ui/Img.svelte';
+	import ImgBanner from '$lib/components/ui/ImgBanner.svelte';
+	import { i18n } from '$lib/stores/i18n.store';
+	import { modalStore } from '$lib/stores/modal.store';
+	import type { DappDescription } from '$lib/types/dappDescription';
+	import { replacePlaceholders } from '$lib/utils/i18n.utils';
+
+	export let dAppDescription: DappDescription;
+	const formattedUrl = new URL(dAppDescription.website).hostname.replace('www.', '');
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:17]: 1. This can crash an error

2. Why removing `www`? 

3. nitpick: for coding style consistency prefer destructing objects

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+</script>
+
+<Modal on:nnsClose={modalStore.close}>
+	<div slot="title">
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:21]: The title should be a fragment

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		<span class="text-center text-xl">{dAppDescription.name}</span>
+	</div>
+
+	<div class="stretch flex flex-col gap-4 pt-4">
+		{#if nonNullish(dAppDescription.screenshots) && dAppDescription.screenshots.length > 0}
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:26]: Isn't this the same piece of code as the one we reviewed recently? Shouldn't you create a component for it?

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+			<div class="overflow-hidden rounded-3xl">
+				<ImgBanner
+					src={dAppDescription.screenshots[0]}
+					alt={replacePlaceholders($i18n.dapps.alt.website, { $dAppname: dAppDescription.name })}
+				/>
+			</div>
+		{/if}
+
+		<div class="rounded-3xl p-5 shadow">
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:35]: Should this be an article?

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			<div class="flex items-center gap-4 border-b border-light-grey pb-5">
+				<div class="h-10 w-10">
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:37]: Why not providing a size to the image to spare a div?

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+					<Img
+						src={dAppDescription.logo}
+						alt={replacePlaceholders($i18n.dapps.alt.logo, { $dAppname: dAppDescription.name })}
+					/>
+				</div>
+				<div>
+					<div class="text-lg font-bold">{dAppDescription.name}</div>
+					<div class="text-sm text-misty-rose">{formattedUrl}</div>
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:45]: This should be a link I guess?

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+				</div>
+				<div class="ml-auto flex space-x-4">
+					{#if dAppDescription.twitter}
+						<ExternalLinkIcon
+							href={dAppDescription.twitter}
+							ariaLabel={$i18n.dapps.alt.open_twitter}
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:51]: There isn't a placeholder in the label that should be replaced with the dapp name?

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+						>
+							<IconTwitter />
+						</ExternalLinkIcon>
+					{/if}
+					{#if dAppDescription.github}
+						<ExternalLinkIcon
+							href={dAppDescription.github}
+							ariaLabel={$i18n.dapps.alt.source_code_on_github}
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:59]: There isn't a placeholder in the label that should be replaced with the dapp name?

File: src/frontend/src/lib/components/dapps/DappTags.svelte [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<script lang="ts">
+	import { i18n } from '$lib/stores/i18n.store';
+	import { replacePlaceholders } from '$lib/utils/i18n.utils';
+
+	export let tags: string[] = [];
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappTags.svelte:5]: Are you going to reuse this component elsewhere? If not, should we pass the dapp description and destructure the tags and dappme within the component to simplify its interface?

File: src/frontend/src/lib/components/ui/ExternalLink.svelte [hunk 10]
```diff
@@ -14,7 +14,7 @@
 	{href}
 	rel="external noopener noreferrer"
 	target="_blank"
-	class="inline-flex items-center gap-2 no-underline"
+	class="inline-flex items-center gap-2 no-underline {$$restProps.class || ''}"
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/ui/ExternalLink.svelte:17]: Shouldn't we provide this in another PR? Seems unrelated.

File: src/frontend/src/lib/stores/modal.store.ts [hunk 11]
```diff
@@ -97,6 +99,7 @@ const initModalStore = <T>(): ModalStore<T> => {
 		openReceiveBitcoin: () => set({ type: 'receive-bitcoin' }),
 		openAboutWhat: () => set({ type: 'about-what' }),
 		openAboutHow: () => set({ type: 'about-how' }),
+		openDappDetails: () => set({ type: 'dapp-details' }),
```
Comment (peterpeterparker) [inline src/frontend/src/lib/stores/modal.store.ts:102]: Shouldn't opening the modal also receive the data as parameter - i.e. the dapp description?

File: src/frontend/src/lib/components/dapps/DappTags.svelte (hunk 9 above)
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappTags.svelte:5]: I'm going to reuse it

File: src/frontend/src/lib/components/ui/ExternalLink.svelte (hunk 10 above)
Comment (loki344) [reply src/frontend/src/lib/components/ui/ExternalLink.svelte:17]: In my opinion not - This change is necessary to implement this modal. But let me know if you want a separate PR and I will make one.

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 1 above)
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte:17]: 1. I see your point, we validate the data with zod to be a valid URL. So we would expect someone to use this component with other data than provided by the json in order to provide an invalid URL. I can handle this case gracefully by not rendering any links to this app - Which makes this modal a pure info modal, but I think this is fine for this edge case.
2. matching the design
3. done

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 3 above)
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte:26]: The promobanner expects a screenshot. This modal can display dapps without screenshots. The promobanner needs position absolute etc. I think the abstraction with the ImgBanner is good enough.

File: src/frontend/src/lib/components/ui/ExternalLink.svelte (hunk 10 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/components/ui/ExternalLink.svelte:17]: on a separate note, should there be `nonNullish` here or `??` coalescing? I thought the linter would spot the second one...

File: src/frontend/src/lib/stores/modal.store.ts (hunk 11 above)
Comment (loki344) [reply src/frontend/src/lib/stores/modal.store.ts:102]: Oh I like this, did not catch this pattern. Trying to figure out how to get the data in the modal if it's passed like this.

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	export let dAppDescription: DappDescription;
+	const formattedUrl = new URL(dAppDescription.website).hostname.replace('www.', '');
+</script>
+
+<Modal on:nnsClose={modalStore.close}>
+	<div slot="title">
+		<span class="text-center text-xl">{dAppDescription.name}</span>
+	</div>
+
+	<div class="stretch flex flex-col gap-4 pt-4">
```
[the remark this answers is not in the document]
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte:38]: I would have to add an extension point to pass in the classname. Also, I'm not really sure if my content here has a 'toolbar' in that sense

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 1 above)
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: made it reactive

**Comment by loki344** [comment]
rdy for second round

File: src/frontend/src/lib/components/dapps/DappTags.svelte (hunk 9 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappTags.svelte:5]: Oki doki, let's see then. Thanks for the feedback!

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 1 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: 1. We don't have tests in place for either the types using Zod or this component. Therefore no certainty that this won't break someday. Please handle any potential errors.

2. Can you please review this with the designer?

- I assume that if a dapp has chosen to use www, they most likely want to communicate using that URL.
- If we decide to ignore their preferred URL, but a non-www URL is available, we can configure this at the metadata level.
- Given the critical importance of the hostname on the IC, particularly with respect to Internet Identity, I believe we should avoid altering the URL of projects.

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 3 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: Oki doki

File: src/frontend/src/lib/components/ui/ExternalLink.svelte (hunk 10 above)
Comment (peterpeterparker) [reply src/frontend/src/lib/components/ui/ExternalLink.svelte]: I think this should have been a separate PR because we don't spread classes but use a property named `styleClass` when passing a class to a component. However, if you want to keep this in this PR, please adjust the code accordingly for consistency reasons.

File: src/frontend/src/lib/components/dapps/DappModal.svelte (hunk 1 above)
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: It's only about the visual appearance of the url, the link in the href is not touched by this formatting. I already checked with the designer: "Cool, let's then make it different. I was interested in cleaner look, so no https:// for sure, we can also try and drop www"

Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: "Again, given the importance of hostnames on the IC, I personally would not mess with those, but if that's the wish, then let's not solve it at the code level and instead use a URL without 'www' in the metadata.

Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappModal.svelte]: Hm ok, as he stated 'try and drop www'. I go with your feedback. thanks

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	import { i18n } from '$lib/stores/i18n.store';
+	import { modalStore } from '$lib/stores/modal.store';
+	import type { DappDescription } from '$lib/types/dappDescription';
+	import { replacePlaceholders } from '$lib/utils/i18n.utils';
+
+	export let dAppDescription: DappDescription;
+	$: ({ website, screenshots, twitter, github, tags, name, description, logo } = dAppDescription);
+
+	let isURLValid: boolean;
+	let formattedUrl: string | undefined;
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte]: To spare two potential variables that are not in sync, you can use a type `formattedUrl: OptionString` and set the value to `null` of the parsing fails in error.

We generally use this pattern "undefined === not set, null == we set that explicitely to null" for those use cases.

File: src/frontend/src/lib/components/dapps/DappModal.svelte [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+				</div>
+			</div>
+
+			<p class="m-0 my-5 text-sm">
+				<Html text={description} />
+			</p>
+			<DappTags dAppName={name} {tags} />
+		</article>
+
+		{#if websiteURL}
```
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappModal.svelte:96]: `nonNullish`

Note that we have a custom eslint rule in progress to check those things.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

**Comment by loki344** [comment]
@AntonioVentilii-DFINITY also good for you?

**Review by AntonioVentilii-DFINITY — APPROVED** [review]
LGTM tks
